### PR TITLE
`redundant_slicing` precedence fix

### PR DESCRIPTION
**Pull request by Jarcho**
fixes: #7972
changelog: Fix `redundant_slicing` suggestion when the result is used in a postfix expression

### CODE REVIEW

File: clippy_utils/src/source.rs [hunk 1]
```diff
@@ -306,6 +306,199 @@ pub fn snippet_with_context(
     )
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ExprPosition {
```
Comment (camsteffen) [inline clippy_utils/src/source.rs:310]: Can we use `ExprPrecedence` from rustc?

Comment (Jarcho) [reply clippy_utils/src/source.rs:310]: Doesn't distinguish between the left and right. Given `a - b` replacing `a` with `x + y` doesn't require parenthesis, replacing `b` does.

Comment (camsteffen) [reply clippy_utils/src/source.rs:310]: Here is how [pretty print](https://github.com/rust-lang/rust/blob/b0535508474fb4353a3b83ef79df87817f028a71/compiler/rustc_ast_pretty/src/pprust/state.rs#L1839-L1847) does it. I guess you'd still have to pass `is_rhs: bool` in some form so 🤷.

Comment (xFrednet) [reply clippy_utils/src/source.rs:310]: Seams like there is no ideal solution. I think this way is acceptable, are you also fine with it @camsteffen?

File: clippy_lints/src/redundant_slicing.rs [hunk 2]
```diff
@@ -57,34 +58,47 @@ impl LateLintPass<'_> for RedundantSlicing {
...
-                    format!("{}{}", reborrow_str, snip),
+                    if !reborrow_str.is_empty() && position > ExprPosition::Prefix {
+                        format!("({}{})", reborrow_str, snip)
+                    } else {
+                        format!("{}{}", reborrow_str, snip)
+                    },
                     app,
                 );
             }
+
```
Comment (xFrednet) [inline clippy_lints/src/redundant_slicing.rs:101]:
```suggestion
```

File: clippy_utils/src/source.rs [hunk 3]
```diff
@@ -306,6 +306,199 @@ pub fn snippet_with_context(
...
+/// snippet of the chile of the borrow expression will get a snippet of what `vec![]` expands in to.
+/// With the target context set to the same as the borrow expression, this will get a snippet of the
+/// call to the macro.
+///
+/// The applicability will be modified in two ways:
+/// * If a snippet can't be extracted it will be changed from `MachineApplicable` or
+///   `MaybeIncorrect` to `HasPlaceholders`.
+/// * If the snippet is taken from a macro expansion then it will be changed from
+///   `MachineApplicable` to `MaybeIncorrect`.
+pub fn snippet_expr(
```
Comment (xFrednet) [inline clippy_utils/src/source.rs:359]: The tests don't have any instance where the parenthesis are added by this function. Can you please add some test cases to ensure that this works correctly?

File: clippy_utils/src/source.rs [hunk 4]
```diff
@@ -306,6 +306,199 @@ pub fn snippet_with_context(
...
+
+/// Extracts a snippet of the given expression taking into account the `SyntaxContext` the snippet
+/// needs to be taken from. Parenthesis will be added if needed to place the snippet in the target
+/// precedence level. Returns a placeholder (`(..)`) if a snippet can't be extracted (e.g. an
+/// invalid span).
+///
+/// The `SyntaxContext` of the expression will be walked up to the given target context (usually
+/// from the parent expression) before extracting a snippet. This allows getting the call to a macro
+/// rather than the expression from expanding the macro. e.g. In the expression `&vec![]` taking a
+/// snippet of the chile of the borrow expression will get a snippet of what `vec![]` expands in to.
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_utils/src/source.rs:350]: I swear I know how to type.

File: clippy_utils/src/source.rs (hunk 1 above)
Comment (camsteffen) [reply clippy_utils/src/source.rs:310]: Yeah this works.

**Comment by Jarcho** [comment]
Just btw I'm working on #7986 which is an alternative to the extra util functions.

File: clippy_utils/src/source.rs [hunk 5]
```diff
@@ -306,6 +306,199 @@ pub fn snippet_with_context(
...
+        Cow::Owned(mut snip) => {
+            let ctxt = expr.span.ctxt();
+
+            // Attempt to determine if parenthesis are needed base on the target position. The snippet may have
+            // parenthesis already, so attempt to find those.
+            // TODO: Remove parenthesis if they aren't needed at the target position.
+            let needs_paren = match expr.peel_drop_temps().kind {
+                ExprKind::Binary(_, lhs, rhs)
+                    if (ctxt == lhs.span.ctxt() && expr.span.lo() != lhs.span.lo())
+                        || (ctxt == rhs.span.ctxt() && expr.span.hi() != rhs.span.hi()) =>
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_utils/src/source.rs:380]: It has to do with macro expansions and whether parenthesis are from the current expression or the sub expression. If the span on the current expansion doesn't match with the outer edges of the two sub expressions then the current expression has parenthesis already, but this doesn't work if either sub expression is the result of a macro expansion.

**Comment by xFrednet** [comment]
Hey @Jarcho, are you planning to continue working on this PR or are you waiting on #7986 before doing more work here? :upside_down_face:

**Comment by Jarcho** [comment]
I'll give it a fix for the specific issue as that doesn't need anything added to utils.

**Comment by Jarcho** [comment]
Closing in favour of #8218.

**Pull request closed without merging**